### PR TITLE
Internal: Container Upgrade scripts - Scroll Snap [ED-11634]

### PR DESCRIPTION
**Pull request by TzviRabinovitch**

### CODE REVIEW

File: tests/phpunit/elementor/modules/usage/test-module.php [hunk 1]
```diff
@@ -47,6 +47,65 @@ class Test_Module extends Elementor_Test_Base {
 		],
 	];
 
+	static $document_mock_default_with_container = [
```
Comment (louiswol94) [inline tests/phpunit/elementor/modules/usage/test-module.php:50]: @TzviRabinovitch Can you maybe mock the following structures as well?

- Nested elements
- A page that has two parent containers, and one of them has a child container.

WDYT?

**Review by louiswol94 — CHANGES_REQUESTED** [review]

File: core/upgrade/upgrades.php [hunk 2]
```diff
@@ -885,4 +911,64 @@ private static function notice( $message ) {
...
+	 */
+	private static function convert_to_inner_containers( $data ) {
+		return Plugin::$instance->db->iterate_data(
+			$data, function ( $element ) {
+
+				if ( 'container' !== $element['elType'] || ! isset( $element['elements'] ) ) {
+					return $element;
+				}
+
+				foreach ( $element['elements'] as &$inner_element ) {
```
Comment (louiswol94) [inline core/upgrade/upgrades.php:945]: @TzviRabinovitch Why is this passed by reference? `&$inner_element`

Comment (TzviRabinovitch) [reply core/upgrade/upgrades.php:945]: @louiswol94 
it is being updated 
i test w/o passing by ref and it was not updated

File: core/upgrade/upgrades.php [hunk 3]
```diff
@@ -885,4 +911,64 @@ private static function notice( $message ) {
...
+	private static function save_updated_document( $post_id, $data ) {
+		$document = Plugin::$instance->documents->get( $post_id );
+
+		$document->save(
+			[
+				'elements' => $data,
+			]
+		);
+
+		return $document;
```
Comment (louiswol94) [inline core/upgrade/upgrades.php:971]: @TzviRabinovitch Does this need to be returned?

File: tests/phpunit/elementor/core/upgrade/test-upgrades.php [hunk 4]
```diff
@@ -453,4 +477,4 @@ private function create_image() {
 	private function delete_image( $attachment_id ) {
 		wp_delete_attachment( $attachment_id, true );
 	}
-}
+}
```
Comment (louiswol94) [inline tests/phpunit/elementor/core/upgrade/test-upgrades.php:480]: Blank line?

File: core/upgrade/upgrades.php [hunk 5]
```diff
@@ -885,4 +911,64 @@ private static function notice( $message ) {
...
+					AND `meta_value` LIKE \'%"elType":"' . $element_type . '"%\';'
+		);
+	}
+
+	/**
+	 * @param $data
+	 *
+	 * @return array|mixed
+	 */
+	private static function convert_to_inner_containers( $data ) {
```
Comment (louiswol94) [inline core/upgrade/upgrades.php:937]:
```suggestion
	private static function maybe_convert_to_inner_containers( $data ) {
```
@TzviRabinovitch Also here?

File: core/upgrade/upgrades.php [hunk 6]
```diff
@@ -833,6 +833,32 @@ public static function _v_3_8_0_fix_php8_image_custom_size() {
...
+			}
+
+			if ( empty( $data ) ) {
+				continue;
+			}
+
+			$data = self::maybe_convert_to_inner_containers( $data );
+
+			self::save_updated_document( $post_id, $data );
+
```
Comment (louiswol94) [inline core/upgrade/upgrades.php:858]:
```suggestion
```

File: tests/phpunit/elementor/core/upgrade/test-upgrades.php (hunk 4 above)
Comment (TzviRabinovitch) [reply tests/phpunit/elementor/core/upgrade/test-upgrades.php:526]: @louiswol94 
"phpcs: File must end with a newline character"

Comment (louiswol94) [reply tests/phpunit/elementor/core/upgrade/test-upgrades.php:525]: @TzviRabinovitch Yes I asked you to add the blank line. In the diff it showed you deleted the blank line. Sorry if I was unclear.

File: tests/phpunit/elementor/core/upgrade/test-upgrades.php [hunk 7]
```diff
@@ -430,6 +431,26 @@ public function test_v_3_8_0_fix_php8_image_custom_size() {
...
+		$updater = $this->create_updater();
+
+		$documents[] = $this->create_document_with_data( Test_Module::$document_mock_default_with_container );
+		$documents[] = $this->create_document_with_data( Test_Module::$document_mock_default );
+		$documents[] = $this->create_document_with_data( Test_Module::$document_mock_nested_tabs );
+
+		Upgrades::_v_3_16_0_container_updates( $updater );
+
+		$this->assert_container_changed( $this->get_decoded_post_meta( $documents[0] ) );
+		$this->assert_sections_not_changed( $this->get_decoded_post_meta( $documents[1] ) );
```
Comment (louiswol94) [inline tests/phpunit/elementor/core/upgrade/test-upgrades.php:449]: @TzviRabinovitch Can you maybe use `$document->get_json_meta( '_elementor_data' );` here instead?

File: tests/phpunit/elementor/core/upgrade/test-upgrades.php [hunk 8]
```diff
@@ -453,4 +474,52 @@ private function create_image() {
...
+
+	/**
+	 * @param $documents
+	 *
+	 * @return mixed
+	 */
+	public function get_decoded_post_meta( $documents ) {
+		$post_meta = get_post_meta( $documents->get_main_id(), '_elementor_data' );
+		return json_decode( $post_meta[0] );
+	}
```
Comment (louiswol94) [inline tests/phpunit/elementor/core/upgrade/test-upgrades.php:524]: If you can use` $document->get_json_meta( '_elementor_data' );` then this function can be removed.

File: tests/phpunit/elementor/core/upgrade/test-upgrades.php (hunk 7 above)
Comment (TzviRabinovitch) [reply tests/phpunit/elementor/core/upgrade/test-upgrades.php:449]: doing that throws the same error as in adding the test with the nested tabs 
we will try to fix it in the separate task

**Review by louiswol94 — APPROVED** [review]

**Pull request merged**